### PR TITLE
Fix a few issues regarding the s3 input

### LINKED ISSUE #10: Setting up a prefix in logstash make him treat it like an actual file
If you set up a prefix in the config to target specific logs, `aws-sdk` will treat the prefixed directory as an actual file and the plugin will crash because he will try to treat it as a log file.

We dont have a test for this particular case and the bug was introduced when we added the option to backup in another s3 prefix.

### PR DESCRIPTION
**Pull request by ph**
* Allow the user to define where the plugin should keep his temporary files
* Bugfix, If a S3 object key exactly match the prefix we will skip it fixes #10
* Use info instead of error when using a user defined sincedb_path
* adding integration test
* Move back the GZIP file and the normal file logic back into the plugin.

### CODE REVIEW

**Comment by ph** [comment]
@purbon I wish I had your rspec presentation before writing this. Could review/test?

**Comment by purbon** [comment]
Jajaja, will do!

**Comment by suyograo** [comment]
@purbon review?

**Comment by purbon** [comment]
@suyograo will do, will do, hope to do it today.

**Comment by ph** [comment]
We will need to yank LogStash::Codecs::GzipLines and LogStash::Codecs::Cloudfront and remove the two repositories since I have moved back the functionality into the S3 input.

https://github.com/logstash-plugins/logstash-codec-cloudfront and  https://github.com/logstash-plugins/logstash-codec-gzip_lines

**Comment by purbon** [comment]
@ph reviewing

File: lib/logstash/inputs/s3.rb [hunk 1]
```diff
@@ -134,13 +136,116 @@ def backup_to_dir(filename)
...
+        # The line need to go through the codecs to replace
+        # unknown bytes in the log stream before doing a regexp match or
+        # you will get a `Error: invalid byte sequence in UTF-8'
+        if event_is_metadata?(event)
+          @logger.debug('Event is metadata, updating the current cloudfront metadata', :event => event)
+          update_metadata(metadata, event)
+        else
+          decorate(event)
+
+          event["cloudfront_version"] = metadata[:cloudfront_version] unless metadata[:cloudfront_version].nil?
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: I would avoid as much as possible the usage of unless, is more clean when you use if not. But this is a personal style test.

File: lib/logstash/inputs/s3.rb [hunk 2]
```diff
@@ -134,13 +136,116 @@ def backup_to_dir(filename)
...
+
+  private
+  def event_is_metadata?(event)
+    line = event['message']
+    version_metadata?(line) || fields_metadata?(line)
+  end
+
+  private
+  def version_metadata?(line)
+    line.start_with?('#Version: ')
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: Is the space at the end of the string important? I mean, what happen if the line start with ```#Version:0.3.4```?

File: lib/logstash/inputs/s3.rb [hunk 3]
```diff
@@ -134,13 +136,116 @@ def backup_to_dir(filename)
...
+  end
+
+  private
+  def fields_metadata?(line)
+    line.start_with?('#Fields: ')
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: Same thing here.

File: lib/logstash/inputs/s3.rb [hunk 4]
```diff
@@ -134,13 +136,116 @@ def backup_to_dir(filename)
...
+  end
+
+  private 
+  def update_metadata(metadata, event)
+    line = event['message'].strip
+
+    if version_metadata?(line)
+      metadata[:cloudfront_version] = line.split(/#Version: (.+)/).last
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: Would not be better to use a regexp to get the version and field here? Could be dry this into their own function?

File: logstash-input-s3.gemspec [hunk 5]
```diff
@@ -23,9 +23,10 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'logstash', '>= 1.4.0', '< 2.0.0'
   s.add_runtime_dependency 'logstash-mixin-aws'
   s.add_runtime_dependency 'logstash-codec-line'
-  s.add_runtime_dependency 'aws-sdk'
```
Comment (purbon) [inline logstash-input-s3.gemspec]: :+1:

File: logstash-input-s3.gemspec [hunk 6]
```diff
@@ -23,9 +23,10 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'logstash', '>= 1.4.0', '< 2.0.0'
   s.add_runtime_dependency 'logstash-mixin-aws'
   s.add_runtime_dependency 'logstash-codec-line'
```
Comment (purbon) [inline logstash-input-s3.gemspec]: why do we need this if you changed to plain?

File: spec/fixtures/invalid_utf8.log [hunk 7]
```diff
@@ -0,0 +1,2 @@
+2015-01-01T02:52:45.866722Z no "GET http://www.logstash.com:80/utfmadness/≈4od HTTP/1.1"
+
```
Comment (purbon) [inline spec/fixtures/invalid_utf8.log]: is this new line necessary?

File: spec/inputs/s3_spec.rb [hunk 8]
```diff
@@ -69,6 +69,21 @@
       expect(config.list_new_files).to eq([present_object.key])
     end
 
+    it 'should ignore file if the file match the prefix' do
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: Just personal test, but I find using should in descriptions a bit strange (http://betterspecs.org/#should)

File: spec/inputs/s3_spec.rb [hunk 9]
```diff
@@ -69,6 +69,21 @@
...
+        prefix = 'mysource/'
+
+        objects_list = [
+          double(:key => prefix, :last_modified => Time.now),
+          present_object
+        ]
+
+        allow_any_instance_of(AWS::S3::ObjectCollection).to receive(:with_prefix).with(prefix) { objects_list }
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: Can we mock this out and avoid using allow_any_instance_of?

File: spec/inputs/s3_spec.rb [hunk 10]
```diff
@@ -69,6 +69,21 @@
...
+
+        config = LogStash::Inputs::S3.new(settings.merge({ 'prefix' => prefix }))
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: can we have this as a subject?

Comment (purbon) [inline spec/inputs/s3_spec.rb]: This apply to all the ```Logstash::Input::S3``` reference here, we could then redefine the settings merge somehow.

File: spec/inputs/s3_spec.rb [hunk 11]
```diff
@@ -144,4 +159,58 @@
...
   end
+
+  context 'when working with logs' do
+    let(:objects) { [log] }
+    let(:log) { double(:key => 'uncompressed.log', :last_modified => Time.now - 2 * day) }
+
+    before do
+      allow_any_instance_of(AWS::S3::ObjectCollection).to receive(:with_prefix).with(nil) { objects }
+      allow_any_instance_of(AWS::S3::ObjectCollection).to receive(:[]).with(log.key) { log }
+      expect(log).to receive(:read)  { |&block| block.call(File.read(log_file)) }
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: expect or allow here?

File: spec/inputs/s3_spec.rb [hunk 12]
```diff
@@ -144,4 +159,58 @@
...
+    end
+
+    context 'compressed' do
+      let(:log) { double(:key => 'log.gz', :last_modified => Time.now - 2 * day) }
+      let(:log_file) { File.join('spec', 'fixtures', 'compressed.log.gz') }
+
+      it 'should process events' do
+        events = fetch_events(settings)
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: can we DRY this and use it as a subject most probably?

File: spec/inputs/s3_spec.rb [hunk 13]
```diff
@@ -144,4 +159,58 @@
...
+    context 'cloudfront' do
+      let(:log_file) { File.join('spec', 'fixtures', 'cloudfront.log') }
+
+      it 'should extract metadata from cloudfront log' do
+        events = fetch_events(settings)
+
+        expect(events.size).to eq(2)
+
+        events.each do |event|
+          expect(event['cloudfront_fields']).to eq('date time x-edge-location c-ip x-event sc-bytes x-cf-status x-cf-client-id cs-uri-stem cs-uri-query c-referrer x-page-url​  c-user-agent x-sname x-sname-query x-file-ext x-sid')
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: Can we break this into different test examples?

File: spec/integration/s3_spec.rb [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  
+  let(:minimal_settings)  {  { "access_key_id" => ENV['AWS_ACCESS_KEY_ID'],
+                               "secret_access_key" => ENV['AWS_SECRET_ACCESS_KEY'],
+                               "bucket" => ENV['AWS_LOGSTASH_TEST_BUCKET'],
+                               "region" => ENV["AWS_REGION"] || "us-east-1",
+                               "prefix" => prefix,
+                               "temporary_directory" => temporary_directory } }
+
+  it "support prefix to scope the remote files" do
+    events = fetch_events(minimal_settings)
```
Comment (purbon) [inline spec/integration/s3_spec.rb]: We should try to DRY this call as a subject of this test.

File: spec/integration/s3_spec.rb [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    expect(events.size).to eq(4)
+  end
+
+  it "backup to another directory" do
+    fetch_events(minimal_settings.merge({ "backup_to_bucket" => "logstash-s3-input-backup"}))
+    expect(list_remote_files("", "logstash-s3-input-backup").size).to eq(2)
+    delete_bucket("logstash-s3-input-backup")
```
Comment (purbon) [inline spec/integration/s3_spec.rb]: Is this something that should go into an after(:example) or after block?

File: spec/integration/s3_spec.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  it "allow you to backup to a local directory" do
+    Stud::Temporary.directory do |backup_dir|
+      fetch_events(minimal_settings.merge({ "backup_to_dir" => backup_dir }))
+      expect(Dir.glob(File.join(backup_dir, "*")).size).to eq(2)
+    end
+  end
+end
+
+# delete_files(prefix)
+def upload_file(local_file, remote_name)
```
Comment (purbon) [inline spec/integration/s3_spec.rb]: This should in a helper/utility file.

File: lib/logstash/inputs/s3.rb (hunk 2 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: the space is important, this is defined in the cloudfront format.

File: lib/logstash/inputs/s3.rb (hunk 3 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: Same as before

File: spec/integration/s3_spec.rb (hunk 16 above)
Comment (ph) [inline spec/integration/s3_spec.rb]: agree

**Comment by ph** [comment]
@purbon  I did not applied all the refactoring on the spec, this PR fix a few important issues with this plugin. I have created a new ticket #19 to do the refactoring of the test suite.

**Pull request closed without merging**